### ISSUE TITLE
Paperclip doesn't understand recursive structs

### ISSUE THREAD

**Issue by izik1** [wafflespeanut/paperclip#276]
```rust
#[derive(paperclip::actix::Apiv2Schema)]
pub struct Foo {
    pub foos: Vec<Foo>,
}
```
generates:
```rust
pub struct Foo {
    pub foos: Vec<Foo>,
}
impl paperclip::v2::schema::Apiv2Schema for Foo {
    const NAME: Option<&'static str> = Some("Foo");
    const DESCRIPTION: &'static str = "";
    fn raw_schema() -> paperclip::v2::models::DefaultSchemaRaw {
        use paperclip::v2::models::{DataType, DataTypeFormat, DefaultSchemaRaw};
        use paperclip::v2::schema::TypedData;
        let mut schema = DefaultSchemaRaw {
            name: Some("Foo".into()),
            ..Default::default()
        };
        schema.data_type = Some(DataType::Object);
        {
            let mut s = Vec::<Foo>::raw_schema();
            if !"".is_empty() {
                s.description = Some("".to_string());
            }
            schema.properties.insert("foos".into(), s.into());
        }
        if Vec::<Foo>::REQUIRED {
            schema.required.insert("foos".into());
        }
        schema
    }
}
impl paperclip::actix::OperationModifier for Foo {}
```
Which will cause a stack overflow at runtime (because `Foo::raw_schema()` calls `Vec<Foo>::raw_schema()` calls...)

I'd also appreciate a workaround if at all possible :bowing_woman:

**Comment by ManevilleF** [comment]
Rust doesn't allow recursive structs without a `Box`.

Yor example can't compile unless your struct looks like:
```rust
pub struct Foo {
    pub foos: Vec<Box<Foo>>,
}
```

Which cause a `paperclip` compilation  error  because it doesn't handle `Box<T>`

**Comment by izik1** [comment]
No... Well, Rust won't allow for recursive structs without _indirection_ because of infinite size, however, a `Vec<T>` already provides the indirection, and has a size on the stack of precisely 24 bytes on x86-64, regardless of `T`.
In fact, the example I provided _compiles_ just fine, it fails at runtime due to infinite recursion, and adding a box won't help with that

**Comment by tiagolobocastro** [comment]
This might be tricky. For each field of the struct  we'd need to check if there's a recursion...
I wonder if we can use the ref link without using accessing raw_schema and thus avoiding this issue altogether.
When I have some time I'll investigate.

**Comment by ManevilleF** [comment]
> No... Well, Rust won't allow for recursive structs without _indirection_ because of infinite size, however, a `Vec<T>` already provides the indirection, and has a size on the stack of precisely 24 bytes on x86-64, regardless of `T`.
> In fact, the example I provided _compiles_ just fine, it fails at runtime due to infinite recursion, and adding a box won't help with that

Wow you are right, I was positive this couldn't work but indeed `Vec<>` `Hawhmap<>` and other containers already allocate on the heap. My apologies

**Comment by ManevilleF** [comment]
> This might be tricky. For each field of the struct we'd need to check if there's a recursion...
> I wonder if we can use the ref link without using accessing raw_schema and thus avoiding this issue altogether.
> When I have some time I'll investigate.

When parsing checking for self reference just as checking for a different struct deriving `Apiv2Schema` isn't possible?

**Comment by tiagolobocastro** [comment]
So at the moment the spec definitions are added without any `$ref` because we need to add the type definitions to the definitions part of the spec whenever we add a service handler using the `Mountable` trait. So, if we have a handler which returns the type `Type1`:
`struct Type1 { field1: Type2, field2: Vec<Type3> }`, we add the schema for `Type`, which then adds the schema for `Type2` and `Vec<Type3>` which then adds the schema for `Type3`. We can't use "$refs" for the inner types, because otherwise they would not be added to the spec... but, if Vec<Type3> was Vec<Type1>, we get stuck...
Other than changing `raw_schema` to include the "parent fields" so we can detect the loop, I'm not sure what else we could do...

**Comment by forest1102** [comment]
Any Update on this topic? Or at least I want to know how to ignore the nested field from parsing.

**Comment by drmick** [comment]
> Any Update on this topic? Or at least I want to know how to ignore the nested field from parsing.

I make excluded routes to avoid this problem. They are initialized to .wrap_api and are not included in the paperclip
```
App::new()
   .configure(excluded_routes)
   .wrap_api()
   .configure(routes)
```

**Comment by drmick** [comment]
> Any Update on this topic? Or at least I want to know how to ignore the nested field from parsing.

To avoid this problem I make:

1. Special type for recursive structures

```
use paperclip::v2::models::{DataType, DataTypeFormat, DefaultSchemaRaw};
use paperclip::v2::schema::Apiv2Schema;

#[derive(Deserialize, Debug, Serialize, Clone)]
pub struct RecursiveTypeWrapper<T>(pub T);

impl <T>Apiv2Schema for RecursiveTypeWrapper<T> {
    fn required() -> bool {
        true
    }

    fn raw_schema() -> DefaultSchemaRaw {
        DefaultSchemaRaw {
            data_type: Some(DataType::Array),
            format: Some(DataTypeFormat::Other),
            ..Default::default()
        }
    }
}
```

2. Use this type at controller

```
#[derive(Debug, Serialize, Deserialize, Apiv2Schema)]
pub struct Tree {
    data: RecursiveTypeWrapper<Vec<SomeRecursiveStruct>>,
}

#[api_v2_operation]
pub async fn tree(
    service: web::Data<SomeService>,
) -> Result<Json<Tree>, AppError> {
    let recursive_data: Vec<SomeRecursiveStruct> = service
        .get_recursive_data()
        .await?;

    let wrapper = RecursiveTypeWrapper(recursive_data);
    let response = Tree{data: wrapper};
    Ok(Json(response))
}
```
